Add cumsum and cumprod ops

ibab opened the pull request:
This implements `tf.scan_sum` (cumulative sum) and `tf.scan_prod` (cumulative product) ops.
Unfortunately, this won't compile with the current version of eigen, as I have to make some small changes to `TensorScanOp.h`
But this should be useful to discuss how the API should look and whether we want to have inclusive/exclusive ops, and so on.

I'll also add tests and gradients to the two ops.

This relates to #813.

ibab committed:
Add cumsum and cumprod ops

This adds cumulative sum (`tf.cumsum`) and cumulative product
(`tf.cumprod`) operators based on a new `scan` method in the Eigen
tensor library. A `reverse` kwarg has been implemented for performing
the scan in the reverse direction.

This includes
 - Python wrappers
 - documentation
 - gradient functions
 - tests

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -1447,4 +1447,52 @@ b: Another tensor, of same type and shape as `a`.
 
+// --------------------------------------------------------------------------
+
+REGISTER_OP("ScanSum")
```
Would cumsum/cumprod be more equivalent to numpy / matlab naming?

ibab replied:
I decided to go with `scan_sum` and `scan_prod` based on this comment: https://github.com/tensorflow/tensorflow/issues/813#issuecomment-179365203
The idea was to keep it consistent with `reduce_*`.
But if the consensus is to go with `cumsum`/`cumprod` I'd be happy to change it.

ibab commented:
Here is the corresponding PR in Eigen, which needs to propagate into TensorFlow first before this PR can compile: https://bitbucket.org/eigen/eigen/pull-requests/195/add-small-fixes-to-tensorscanop

Mistobaan commented on tensorflow/core/kernels/scan_ops_common.h:
```diff
@@ -0,0 +1,106 @@
+      HANDLE_SCAN(8);
+      default:
+        OP_REQUIRES(ctx, false,
+                    errors::InvalidArgument("Scan only supports tensors with "
```
what about: `Scan does not support tensors with more than 8 dimensions`

ibab replied:
Thanks! I've changed the message accordingly.

Mistobaan commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -1447,4 +1447,52 @@ b: Another tensor, of same type and shape as `a`.
+    .Output("out: T")
+    .Attr("T: numbertype")
+    .Doc(R"doc(
+Compute the cumulative sum of the input tensor `x` along the axis `axis`.
```
I agree with @vrv `cumsum` is easier to remember and understand what it does. Your description also confirms that. To describe the operation you don't use 'scan' anywhere.

ibab replied:
I've now changed the op names to `cumsum` and `cumprod` 👍

ibab commented:
Seeing as `tf.reverse(tf.scan_X(tf.reverse(x)))` is useful in quite a few situations, would it make sense to add a `reverse=True/False` kwarg to the scan ops?
This could either be done in the Python wrappers, or in C++.

Also, I haven't yet figured out how to write the gradient for `scan_prod` without using the same division trick that's also been used in #2641.

ibab commented:
I've made a few changes:
 - Renamed the ops to `cumsum` and `cumprod`
 - Added GPU support
 - Added a `reverse` kwarg to efficiently compute backwards scans
 - Added gradients (the `cumprod` gradient still uses the `/ x` hack)
 - Added tests

My Eigen PR has also been merged, so we will be able to test the PR as soon as Eigen is updated to the newest version.

girving commented:
Sorry for the delay.  Looking over this now.

girving commented on tensorflow/core/kernels/scan_ops.cc:
```diff
@@ -0,0 +1,175 @@
+                errors::InvalidArgument("ScanOp: axis must be a scalar, not ",
+                                        tensor_axis.shape().DebugString()));
+
+    int axis = tensor_axis.tensor<int, 0>()();
```
Do the following in order to avoid irritating race condition technicalities:

    const int axis = internal::SubtleMustCopy(tensor_axis.scalar<int>()());

We try to avoid any possibility of crashing TensorFlow in cases of asynchronous race conditions.

girving commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -1447,4 +1447,56 @@ b: Another tensor, of same type and shape as `a`.
+
+By setting the `reverse` operation to `True`, the sum is performed in the
+reverse order. In contrast to using `tf.reverse`, this avoids copying the
+tensor.
```
Despite numpy convention, typically when I do a scan I want the exclusive version with `output[0] = identity`.  In particular, that's the kind of scan necessary to do safe gradients of `reduce_prod`.

How difficult would a possibly optional exclusive scan be to arrange?

ibab replied:
I've though about this when writing the op.
I'm not sure how to neatly make the exclusive version accessible in Eigen, except by creating a new `cumsum_exclusive(...)` method.
Maybe it's enough to only add this on the TensorFlow side by optionally padding the tensor with the value that we get from `Reducer::initialize()` ?

girving replied:
Let's get this one in and worry about the exclusive version later if necessary.

ibab replied:
We won't be able to compile this PR right now, as Eigen hasn't been updated to the latest version yet :(
I'll have a look at implementing the `exclusive=True/False` kwarg

girving commented on tensorflow/core/ops/math_ops.cc:
````diff
@@ -1447,4 +1447,56 @@ b: Another tensor, of same type and shape as `a`.
+```prettyprint
+# tensor 'x' is [1, 2, 3, 4, 5]
+tf.cumsum(x)               ==> [1, 3, 6, 10, 15]
+tf.cumsum(x, reverse=True) ==> [5, 9, 12, 14, 15]
````
Is that right?  I would have assumed `[15, 14, 12, 9, 5]`?

girving replied:
We should have doc lines for the inputs and outputs: `x`, `axis`, `reverse`, and `out`.

ibab replied:
Are those not generated automatically when creating the Python functions in `gen_math_ops`?
I remember this because I copied the docstrings of the generated ones into the Python wrappers.
But I'll be happy to add these if there's a way to avoid duplication in the generated docstrings.

girving replied:
If all the information about them is already there in the prose section above, and the autogenerated docs are fine, it's okay to leave as is.

ibab replied:
The only thing that I added are the default values (like `default: 0`).
But these don't apply to the autogenerated ops, so I think it might be best to leave things as they are now.

bryongloden replied:
Yup @vrv

samjabrahams replied:
@girving I believe the `reverse=True` output is correct. It cumulatively adds up the tensor starting from the end and iterates backwards, but it doesn't reverse the output tensor:

```
cumsum([1, 2, 3, 4, 5])
= [(1), (1 + 2), (1 + 2 + 3), (1 + 2 + 3 + 4), (1 + 2 + 3 + 4 + 5)]
= [1, 3, 6, 10, 15]

cumsum([1, 2, 3, 4, 5], reverse=True)
= [(5), (5 + 4), (5 + 4 + 3), (5 + 4 + 3 + 2), (5 + 4 + 3 + 2 +1)] 
= [5, 9, 12, 14, 15]
```

ibab replied:
@samjabrahams 
It's implemented right now by reversing, calculating `cumsum` and reversing again, so:
```
initial:
  [1, 2, 3, 4, 5]
reverse:
  [5, 4, 3, 2, 1]
cumsum:
  [5, 9, 12, 14, 15]
reverse:
  [15, 14, 12, 9, 5]
```
This mainly exists to simplify the gradients, and can also be used when fixing the gradient for `reduce_prod`.

samjabrahams replied:
I see- very cool stuff. My gut wants to suggest splitting it up into separate `reverse_sum` and `reverse_order` parameters, but I don't want to over-complicate things if it's unnecessary.

If you think that change would be beneficial, I'd be happy to take a crack at coding it; I don't want to burden you with the annoying overhead a change like this would cause. If this change is going to happen, it'd be smart to do this now to avoid backwards compatibility issues.

ibab replied:
I'm slightly against making the reverse option a separate op, as things would get complicated once we introduce an `exclusive` option.
(Leading to  `reverse_exclusive_cumsum` if we continue on the same path)

samjabrahams replied:
I don't think it would require an entirely separate Op- I'm just suggesting splitting up the `reverse` parameter into two distinct reversal options. Right now it indicates that it should both reverse the order of the summation (starting from the back) as well as reverse the order of the output tensor.

My thought is to add two more cases in [this section](https://github.com/tensorflow/tensorflow/pull/2711/files#diff-b6fb1bb5fdce94575b577ae6c5082e2fR32) (one each for just `reverse_sum == true` and just `reverse_order == true`) as well as changing `reverse` into those two separate parameters. The current `if (reverse)` would be converted to `if (reverse_sum && reverse_order)`

ibab replied:
Sorry, I misunderstood your suggestion.
I think that if you simply want to reverse the order of the output (or input), using `tf.reverse` would be much more appropriate.

girving replied:
Let's not split the parameters.  The existing parameter makes intuitive sense: it's whether you accumulate from the left or the right.  If you want something else, use `tf.reverse` directly.

girving commented on tensorflow/python/kernel_tests/scan_ops_test.py:
```diff
@@ -0,0 +1,219 @@
+
+from tensorflow.python.framework import errors
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.ops import math_ops
```
Use `tf.` rather than individual imports of `errors`, `tensor_shape`, and `math_ops`.

girving commented on tensorflow/python/kernel_tests/scan_ops_test.py:
```diff
@@ -0,0 +1,219 @@
+    with self.assertRaisesWithPredicateMatch(
+        errors.InvalidArgumentError,
+        lambda e: "Expected scan axis in the range" in str(e)):
+      with self.test_session():
```
Flip the order of the two withs.

girving commented on tensorflow/python/ops/math_grad.py:
```diff
@@ -815,3 +815,22 @@ def _CrossGrad(op, grad):
+def _CumsumGrad(op, grad):
+  axis = op.inputs[1]
+  reverse = op.get_attr("reverse")
+  return [math_ops.cumsum(grad, axis=axis, reverse=(not reverse)), None]
```
That is a lovely gradient function. :)

bryongloden replied:
@ibab The best. Your going to have to explain this to us at RSA 2017

girving commented on tensorflow/python/ops/math_ops.py:
````diff
@@ -1427,6 +1435,74 @@ def tanh(x, name=None):
+    ```prettyprint
+    # tensor 'x' is [1, 2, 3, 4, 5]
+    tf.cumsum(x)               ==> [1, 3, 6, 10, 15]
+    tf.cumsum(x, reverse=True) ==> [5, 9, 12, 14, 15]
````
Is that backwards?

ibab replied:
Yup, it's the same mistake as above.
Thanks for spotting this!

girving commented on tensorflow/python/ops/math_ops.py:
```diff
@@ -1719,6 +1795,12 @@ def _LinspaceShape(op):
 
 
+@ops.RegisterShape("Cumsum")
+@ops.RegisterShape("Cumprod")
```
`@ops.RegisterShape("Cumsum")(common_shapes.unchanged_shape)`

girving commented:
Jenkins, test this please.

ibab commented:
Don't think it'll compile, as the current Eigen version doesn't contain the two fixes that were merged into Eigen two days ago.

girving commented:
@ibab Alas, re the eigen change.  Is there something I should do to help that process along?  If not, can you send me a mention here when we're ready to go?

girving commented:
@benoitsteiner Does something need to happen to get those Eigen changes in?

ibab commented:
We could simply modify the bazel workspace and assorted files to pull the newest Eigen version, but that could interfere with @benoitsteiner's work (might produce a merge/rebase conflict for the affected files).

bryongloden commented:
Or not. Go for it @ibab

ibab commented:
I've added exclusive scan functionality to Eigen here: https://bitbucket.org/eigen/eigen/pull-requests/197/implement-exclusive-scan-option-for-tensor/diff.
I also have a version of `ScanOp` that has an optional `exclusive=True/False` argument, which I can push in a new PR once we finish this one.

ibab commented:
My changes to Eigen have passed in the nightly tests.
@benoitsteiner: Would it be okay to update the Eigen version now?
Is that something I should do as part of this PR?

ibab commented:
@girving: The Eigen version in TensorFlow has been updated, so we can have a shot at testing this.
I've rebased my changes on top of `master` and squashed them into a single commit.
I've also changed the copyright holder from `Google` to `The TensorFlow Authors`.

The GPU test is segfaulting for me, but I've debugged this a bit and it seems to be a problem that affects many of the tests on my machine.
Everything works if I test it manually and force the device with `tf.device('/gpu:0')`, and the CPU tests are working, so I suspect the tests will simply pass if we run them.

girving commented:
Jenkins, test this please.

ibab commented:
Seems like it's also failing on travis.
I can reproduce the segfault on my machine when setting the `opt_level` to `-1` with the following code:
```python
from tensorflow.core.protobuf import config_pb2
config = config_pb2.ConfigProto()
config.graph_options.optimizer_options.opt_level = -1
s = tf.Session(config=config)
with tf.device('/gpu:0'):
  x = np.arange(1, 6).astype(np.float32)
  s.run(tf.cumsum(x, 0, False))
```
Strangely, it works without problems when running with the default opt level.
The segfault occurs when accessing the memory of the input tensor.

girving commented:
Unpleasant!  Probably a job for a debug+optimized compile and valgrind; I don't have any particular smart ideas for what it might be.

benoitsteiner commented:
Can you try with the latest version of Eigen (version [3f47637d1e7c1](https://bitbucket.org/eigen/eigen/commits/3f47637d1e7c199b10c66bf1f8d3bf046e479d26)) ? I just fixed a potential issue in the implementation of the evaluator of the scan operation that may result in dangling references once the evaluator is copied.

ibab commented:
Thanks for fixing the evaluator!
I've compiled with the changes in `3f47637d1e7c1`, but unfortunately the tests are still segfaulting.
I'll try to hunt down the problem using valgrind.

benoitsteiner commented:
@ibab  The problem is that evalSubExprIfNeeded is executed on CPU, while both your input data and the temporary buffer reside on GPU memory. You're going to have to launch a CUDA kernel in which you call actually call accumulateTo for the scan operations to work on GPU.

ibab commented:
Thanks!
I also noticed this when I tried to reproduce the problem with `cuda-gdb`, but didn't actually end up in a kernel.
I had assumed that the kernel launch would be automatically handled by the framework (e.g. in `TensorExecutor`)
I can add the launch in a new PR to eigen.

benoitsteiner commented:
@ibab. TensorExecutor will call the coeff and packet methods from a CUDA kernel. However it will call evalSubExprIfNeeded and cleanup from the host. This makes it possible to do preprocessing on the host if needed. Looking forward to your pull request.

Mistobaan commented:
@ibab Any update on this ? from the thread seems like there is some work to be done in the eigen library before this PR can be successfully merged. Is that correct ?

ibab commented:
@Mistobaan I've made the necessary changes in Eigen here: https://bitbucket.org/eigen/eigen/commits/bbffe8ed5f26dfc7166e9fd97ce15bc2655af769
Currently I'm just waiting for an update of the Eigen version that's used in tensorflow.

ibab commented:
@girving @benoitsteiner Maybe I should update the Eigen version as part of this PR, or in a separate PR?

benoitsteiner commented:
@ibab I'd recommend that you create a separate PR to update Eigen. From time to time we hit a compatibility issue, and we need to update TensorFlow as well. It's best to have a standalone PR that contains both the Eigen update and the corresponding TF changes.

ibab commented:
I've updated the Eigen version, so the tests should pass now (they're passing locally for me).

vrv commented:
@tensorflow-jenkins test this please

vrv commented:
Cool, @girving this looks ready to review.

girving commented:
Looks good.  I read through it again, and nothing seems to have changed since the last review.

@ibab Thank you for the contribution!

ibab commented:
@girving Thanks for reviewing!
